api-compose-object.go

```diff
@@ -0,0 +1,392 @@
+	})
+	defer closeResponse(resp)
+	if err != nil {
+		return p, err
+	}
+
+	// Check if we got an error response.
+	if resp.StatusCode != http.StatusOK {
+		return p, httpRespToErrorResponse(resp, bucket, object)
+	}
+
+	// Decode copy-part response on success.
+	cpObjRes := copyObjectResult{}
+	err = xmlDecoder(resp.Body, &cpObjRes)
+	if err != nil {
+		return p, err
+	}
+	p.PartNumber, p.ETag = partNumber, cpObjRes.ETag
+	return p, nil
+}
+
+// Helper to fetch size and etag of an object using a StatObject call.
+func (c Client) getSourceProps(bucket, object string) (size int64, etag string, err error) {
+	// Get object info - need size and etag here.
+	var objInfo ObjectInfo
+	objInfo, err = c.StatObject(bucket, object)
+	if err != nil {
+		err = fmt.Errorf("Could not stat object - %s/%s: %v", bucket, object, err)
+	} else {
+		size = objInfo.Size
+		etag = objInfo.ETag
+	}
+	return
+}
+
+// ComposeObject - creates an object using server-side copying of
+// existing objects. It takes a list of source objects (with optional
+// offsets) and concatenates them into a new object using only
+// server-side copying operations.
+func (c Client) ComposeObject(dst DestinationInfo, srcs []SourceInfo) error {
+	// Input validation.
+	if err := s3utils.CheckValidBucketName(dst.bucket); err != nil {
+		return err
+	}
+	if err := s3utils.CheckValidObjectName(dst.object); err != nil {
+		return err
+	}
+	if len(srcs) > maxPartsCount {
+		return ErrInvalidArgument("Cannot copy more than 10000 source objects.")
+	}
+
+	// Query and collect source size, and construct
+	// upload-part-copy arguments.
+	type PartInfo struct {
+		start, end int64
+		idx        int
+	}
+	var partInfo []PartInfo
+	var totalSize int64
+	for i, src := range srcs {
+		size, etag, err := c.getSourceProps(src.bucket, src.object)
+		if err != nil {
+			return fmt.Errorf("Could not get source props for %s/%s: %v", src.bucket, src.object, err)
+		}
+
+		// Since we did a HEAD to get size, we use the ETag
+		// value to make sure the object has not changed by
+		// the time we perform the copy. This is done, only if
+		// the user has not set their own ETag match
+		// condition.
+		if src.Headers.Get("x-amz-copy-source-if-match") == "" {
+			src.SetMatchETagCond(etag)
+		}
+
+		// Check if a segment is specified, and if so, is the
+		// segment within object bounds?
+		if src.start != -1 {
+			if src.end >= size {
+				return ErrInvalidArgument(
+					fmt.Sprintf("SourceInfo %d has invalid segment-to-copy [%d, %d] (size is %d)",
+						i, src.start, src.end, size))
+			}
+		}
+
+		// Is data to copy too large?
+		totalSize += size
+		if totalSize > maxMultipartPutObjectSize {
+			return ErrInvalidArgument(fmt.Sprintf("Cannot compose an object of size %d (> 5TiB)", totalSize))
+		}
+
+		// Append copy-part requests info for the current
+		// source object.
+		var currOffset, partSize int64
+		for currOffset < size {
+			if currOffset+copyPartSize < size {
+				partSize = copyPartSize
```

Comment:
I understand your point about the bad case for erasure coding but there is another problem. If we use up more parts for the first few sources, then there may be too many parts for later sources in the input and it may exceed the max parts limit. I will think a bit and figure out a working algo.